File: src/main/java/org/cascadebot/cascadebot/utils/ConfirmUtils.java
/*
 * Copyright (c) 2019 CascadeBot. All rights reserved.
 * Licensed under the MIT license.
 */

package org.cascadebot.cascadebot.utils;

import com.google.common.collect.ArrayListMultimap;
import com.google.common.collect.ListMultimap;
import lombok.experimental.UtilityClass;
import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.entities.TextChannel;
import net.dv8tion.jda.api.requests.ErrorResponse;
import org.cascadebot.cascadebot.CascadeBot;
import org.cascadebot.cascadebot.UnicodeConstants;
import org.cascadebot.cascadebot.data.managers.GuildDataManager;
import org.cascadebot.cascadebot.data.objects.GuildData;
import org.cascadebot.cascadebot.messaging.MessageType;
import org.cascadebot.cascadebot.messaging.Messaging;
import org.cascadebot.cascadebot.tasks.Task;
import org.cascadebot.cascadebot.utils.buttons.Button;
import org.cascadebot.cascadebot.utils.buttons.ButtonGroup;

import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;

@UtilityClass
public class ConfirmUtils {

    // Holds the users that have confirmed an action
    private static ListMultimap<String, ConfirmRunnable> confirmedMap = ArrayListMultimap.create();

    //region Confirm Action
    public static boolean confirmAction(long userId, String actionKey, TextChannel channel, MessageType type, String message, long buttonDelay, long expiry, ConfirmRunnable action) {
        GuildData guildData = GuildDataManager.getGuildData(channel.getGuild().getIdLong());
        boolean useEmbed = guildData.getCoreSettings().isUseEmbedForMessages();
        Message sentMessage;
        try {
            sentMessage = Messaging.sendMessageTypeMessage(channel, type, message, useEmbed).get();
            action.userId = userId;
            action.message = sentMessage;
            confirmedMap.put(actionKey, action);
        } catch (ExecutionException e) {
            return false;
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            return false;
        }
        if (sentMessage == null) {
            return false;
        }

        if (channel.getGuild().getMember(CascadeBot.INS.getSelfUser()).hasPermission(channel, Permission.MESSAGE_ADD_REACTION)) {
            Task.getScheduler().schedule(() -> {
                ButtonGroup group = new ButtonGroup(userId, channel.getIdLong(), channel.getGuild().getIdLong());
                group.addButton(new Button.UnicodeButton(UnicodeConstants.TICK, (runner, channel1, message1) -> {
                    if (runner.getIdLong() != action.userId) return;
                    action.run();
                }));
                group.addButtonsToMessage(sentMessage);
                group.setMessage(sentMessage.getIdLong());
                guildData.addButtonGroup(channel, sentMessage, group);
            }, buttonDelay, TimeUnit.MILLISECONDS);
        }

        Task.getScheduler().schedule(() -> {
            confirmedMap.remove(actionKey, action);
            sentMessage.delete().queue(null, DiscordUtils.handleExpectedErrors(ErrorResponse.UNKNOWN_MESSAGE));
        }, expiry, TimeUnit.MILLISECONDS);

        return true;
    }

    public static void confirmAction(long userId, String actionKey, TextChannel channel, MessageType type, String message, ConfirmRunnable action) {
        confirmAction(
                userId,
                actionKey,
                channel,
                type,
                message,
                TimeUnit.SECONDS.toMillis(2),
                TimeUnit.MINUTES.toMillis(1),
                action
        );
    }

    public static void confirmAction(long userId, String actionKey, TextChannel channel, String message, ConfirmRunnable action) {
        confirmAction(
                userId,
                actionKey,
                channel,
                MessageType.WARNING,
                message,
                TimeUnit.SECONDS.toMillis(2),
                TimeUnit.MINUTES.toMillis(1),
                action
        );
    }
    //endregion

    public static boolean hasConfirmedAction(String actionKey, long userId) {
        return confirmedMap.entries().stream().anyMatch(entry -> entry.getKey().equals(actionKey) && entry.getValue().userId == userId);
    }

    public static void completeAction(String actionKey, long userId) {
        Optional<Map.Entry<String, ConfirmRunnable>> entryOptional = confirmedMap.entries().stream().filter(entry -> entry.getKey().equals(actionKey) && entry.getValue().userId == userId).findFirst();
        entryOptional.ifPresent(stringConfirmRunnableEntry -> stringConfirmRunnableEntry.getValue().run());
    }

    public abstract static class ConfirmRunnable implements Runnable {

        private long userId;
        private Message message;

        @Override
        public void run() {
            message.delete().queue(null, DiscordUtils.handleExpectedErrors(ErrorResponse.UNKNOWN_MESSAGE));
            execute();
        }

        public abstract void execute();

    }

}

File: src/main/kotlin/org/cascadebot/cascadebot/utils/ConfirmUtils.kt
/*
 * Copyright (c) 2020 CascadeBot. All rights reserved.
 * Licensed under the MIT license.
 */

package org.cascadebot.cascadebot.utils

import com.google.common.collect.ArrayListMultimap
import com.google.common.collect.ListMultimap
import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.entities.Emoji
import net.dv8tion.jda.api.entities.Message
import net.dv8tion.jda.api.entities.TextChannel
import net.dv8tion.jda.api.interactions.components.ButtonStyle
import net.dv8tion.jda.api.requests.ErrorResponse
import org.cascadebot.cascadebot.CascadeBot
import org.cascadebot.cascadebot.UnicodeConstants
import org.cascadebot.cascadebot.data.managers.GuildDataManager
import org.cascadebot.cascadebot.messaging.MessageType
import org.cascadebot.cascadebot.messaging.Messaging.sendMessage
import org.cascadebot.cascadebot.tasks.Task
import org.cascadebot.cascadebot.utils.interactions.CascadeActionRow
import org.cascadebot.cascadebot.utils.interactions.CascadeButton
import org.cascadebot.cascadebot.utils.interactions.ComponentContainer
import org.cascadebot.cascadebot.utils.interactions.IButtonRunnable
import java.util.concurrent.ExecutionException
import java.util.concurrent.TimeUnit

private typealias ActionKey = String

object ConfirmUtils {

    // Holds the users that have confirmed an action
    // Map is in the format ActionKey as the key and the list of conformation actions
    @JvmStatic
    private val confirmedMap: ListMultimap<ActionKey, ConfirmationAction> = ArrayListMultimap.create()

    /**
     * Registers an action to be confirmed. This method sends the message specified from the input and
     * adds a ✅ button to it (And also a ❌ button if `isCancellable` is true). If the user clicks the
     * check button before the expiry time, the action is run and the original confimration message is deleted.
     *
     * @param userId The ID of the user this action is associated with. The action can only be run by this user.
     * @param actionKey An internal identified for this action. Should be a short string with lowercase and hyphens
     * only (Although this is not a technical restriction, just a stylistic one)
     * @param channel The channel in which to send the confirmation message.
     * @param type What message type the message should be displayed as. This only affects the color of the message.
     * @param message The confirmation message to send to the user.
     * @param buttonDelay How long between the message sending and the buttons being added. This is useful if you
     * want to make sure the user has read the confirmation.
     * @param expiry How long the confirmation action is valid for. After this length of time, the action will not be run.
     * @param isCancellable Whether the ❌ button will appear to cancel the message.
     * @param action The action to run **only** if the user confirms the action before the expiry time is up.
     *
     * @return Whether the confirmation action has been successfully registered.
     */
    @JvmStatic
    @JvmOverloads
    fun registerForConfirmation(
            userId: Long,
            actionKey: String,
            channel: TextChannel,
            type: MessageType = MessageType.WARNING,
            message: String,
            buttonDelay: Long = TimeUnit.SECONDS.toMillis(2),
            expiry: Long = TimeUnit.MINUTES.toMillis(1),
            isCancellable: Boolean,
            action: Runnable
    ): Boolean {
        //val guildData = GuildDataManager.getGuildData(channel.guild.idLong)
        val confirmationAction: ConfirmationAction
        //val useEmbed = guildData.core.useEmbedForMessages
        val sentMessage: Message
        try {
            sentMessage = sendMessage(type, channel, message, true).get()
            confirmationAction = ConfirmationAction(userId, sentMessage, action)
            confirmedMap.put(actionKey, confirmationAction)
        } catch (e: ExecutionException) {
            return false
        } catch (e: InterruptedException) {
            Thread.currentThread().interrupt()
            return false
        }
        if (sentMessage == null) {
            return false
        }
        if (channel.guild.getMember(CascadeBot.INS.selfUser)!!.hasPermission(channel, Permission.MESSAGE_ADD_REACTION)) {
            Task.getScheduler().schedule({
                val container = ComponentContainer()
                val row = CascadeActionRow()
                container.addRow(row)

                row.addComponent(CascadeButton.success(Emoji.fromUnicode(UnicodeConstants.TICK), IButtonRunnable { member, _, _ ->
                    if (member.idLong != confirmationAction.userId)
                        confirmationAction.run()
                }))

                if (isCancellable) {
                    row.addComponent(CascadeButton.danger(Emoji.fromUnicode(UnicodeConstants.RED_CROSS), IButtonRunnable { _, _, _ ->
                        confirmedMap.remove(actionKey, confirmationAction)
                        sentMessage.delete()
                            .queue(null, DiscordUtils.handleExpectedErrors(ErrorResponse.UNKNOWN_MESSAGE))
                    }))
                }

                sentMessage.editMessageComponents(container.getComponents().map { it.toDiscordActionRow() }).override(true).queue()

                //guildData.addComponents(channel, sentMessage, container)
            }, buttonDelay, TimeUnit.MILLISECONDS)
        }
        Task.getScheduler().schedule({
            confirmedMap.remove(actionKey, confirmationAction)
            sentMessage.delete().queue(null, DiscordUtils.handleExpectedErrors(ErrorResponse.UNKNOWN_MESSAGE))
        }, expiry, TimeUnit.MILLISECONDS)
        return true
    }

    /**
     * Checks whether a user has an action registered under the specified key.
     *
     * @param actionKey The key of the action to check for.
     * @param userId The ID of the user for which we are searching for actions.
     * @return Whether an action matching the criteria is found in the internal map.
     */
    @JvmStatic
    fun hasRegisteredAction(actionKey: String, userId: Long): Boolean {
        return confirmedMap.entries().any { it.key == actionKey && it.value.userId == userId }
    }

    /**
     * Manually confirms an action the user has registered. Useful for when reactions are not enabled in a guild.
     *
     * If multiple actions exist for this user, only the first is run.
     *
     * If no actions exist, this method returns false.
     *
     * @param actionKey The action key for the action to confirm.
     * @param userId The user associated with the action.
     * @return `true` if the action has been run, `false` is no action exists with this criteria.
     */
    @JvmStatic
    fun confirmAction(actionKey: String, userId: Long): Boolean {
        val entryOptional = confirmedMap.entries().firstOrNull { it.key == actionKey && it.value.userId == userId }
        return if (entryOptional != null) {
            entryOptional.value.run()
            true
        } else {
            false
        }
    }

    private class ConfirmationAction(val userId: Long, val message: Message, private val runnable: Runnable) {
        fun run() {
            message.delete().queue(null, DiscordUtils.handleExpectedErrors(ErrorResponse.UNKNOWN_MESSAGE))
            runnable.run()
        }
    }
}